File: Java-version/LeetCode/src/com/jooc/Common/Solution_208.java
package com.jooc.Common;

import com.jooc.Solution;

public class Solution_208 implements Solution {
    @Override
    public void main() {

    }

    class TrieNode{
        private TrieNode[] links;

        private final int capacity = 26;

        private boolean isEnd;

        public TrieNode(){
            links = new TrieNode[capacity];
            isEnd = false;
        }

        public boolean contaisKey(char ch){
            return links[ch - 'a'] != null;
        }

        public TrieNode get(char ch){
            return links[ch - 'a'];
        }

        public void put(char ch, TrieNode node){
            links[ch - 'a'] = node;
        }

        public void setEnd(){
            isEnd = true;
        }

        public boolean isEnd(){
            return isEnd;
        }
    }

    class Trie {

        TrieNode root;

        /** Initialize your data structure here. */
        public Trie() {
            root = new TrieNode();
        }

        /** Inserts a word into the trie. */
        public void insert(String word) {
            TrieNode node = root;
            for(int i = 0; i < word.length(); i++){
                char ch = word.charAt(i);
                if(!node.contaisKey(ch)){
                    node.put(ch, new TrieNode());
                }
                node = node.get(ch);
            }
            node.setEnd();
        }

        private TrieNode searchPrefix(TrieNode root, String prefix){
            TrieNode node = root;
            for(int i = 0; i < prefix.length(); i++){
                char ch = prefix.charAt(i);
                if(node.contaisKey(ch)){
                    node = node.get(ch);
                }else{
                    return null;
                }
            }
            return node;
        }

        /** Returns if the word is in the trie. */
        public boolean search(String word) {
            TrieNode node = searchPrefix(root, word);
            return node != null && node.isEnd();
        }

        /** Returns if there is any word in the trie that starts with the given prefix. */
        public boolean startsWith(String prefix) {
            return searchPrefix(root, prefix) != null;
        }
    }




}

File: python-version/leetcode/Solution_208.py
class Trie:

    def __init__(self):
        self.children = [None] * 26
        self.is_end = False

    def insert(self, word: str) -> None:
        node = self
        for c in word:
            c_index = ord(c) - ord('a')
            if not node.children[c_index]:
                node.children[c_index] = Trie()
            node = node.children[c_index]
        node.is_end = True

    def search_prefix(self, prefix):
        node = self
        for ch in prefix:
            c_index = ord(ch) - ord('a')
            if not node.children[c_index]:
                return None
            node = node.children[c_index]
        return node

    def search(self, word: str) -> bool:
        node = self.search_prefix(word)
        return node is not None and node.is_end

    def startsWith(self, prefix: str) -> bool:
        return self.search_prefix(prefix) is not None



# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)